kdc/httpkadmind.c

```diff
@@ -0,0 +1,2025 @@
+    if (retname && !strncmp("KRB5KDC_", retname, sizeof("KRB5KDC_") - 1))
+        retname += sizeof("KRB5KDC_") - 1;
+#undef PREFIX
+    heim_audit_trail((heim_svc_req_desc)r, ret, retname);
+}
+
+static krb5_log_facility *logfac;
+static pthread_key_t k5ctx;
+
+static krb5_error_code
+get_krb5_context(krb5_context *contextp)
+{
+    krb5_error_code ret;
+
+    if ((*contextp = pthread_getspecific(k5ctx)))
+        return 0;
+
+    ret = krb5_init_context(contextp);
+    /* XXX krb5_set_log_dest(), warn_dest, debug_dest */
+    if (ret == 0)
+        (void) pthread_setspecific(k5ctx, *contextp);
+    return ret;
+}
+
+static int port = -1;
+static int help_flag;
+static int daemonize;
+static int daemon_child_fd = -1;
+static int local_read_only;
+static int verbose_counter;
+static int version_flag;
+static int reverse_proxied_flag;
+static int thread_per_client_flag;
+struct getarg_strings audiences;
+static const char *cert_file;
+static const char *priv_key_file;
+static const char *cache_dir;
+static const char *realm;
+static const char *hdb;
+static const char *primary_server;
+static const char *admin_server;
+static const char *primary_admin_server;
+static const char *stash_file;
+static const char *kadmin_client_name = "keytab_service";
+static const char *kadmin_client_keytab;
+
+static krb5_error_code
+get_kadm_handle(krb5_context context, void **kadm_handle, int primary)
+{
+    kadm5_config_params conf;
+    krb5_error_code ret = 0;
+    const char *pstr;
+    const char *server = NULL;
+    char *s = NULL;
+
+    /* Configure kadmin connection */
+    memset(&conf, 0, sizeof(conf));
+    if (primary && local_read_only && !primary_admin_server)
+        return KADM5_READ_ONLY;
+    if (primary || !admin_server)
+        server = primary_admin_server;
+    else
+        server = admin_server;
+    if (server && (pstr = strrchr(server, ':'))) {
+        int32_t n;
+        char *ends;
+
+        if ((s = strdup(server)) == NULL)
+            return krb5_enomem(context);
+
+        errno = 0;
+        s[pstr - server] = '\0';
+        if ((n = strtol(++pstr, &ends, 10)) < 0 ||
+            n > UINT16_MAX || errno || *ends != '\0') {
+            errno = errno ? errno : ERANGE;
+            err(1, "Port number invalid in: %s", server);
+            free(s);
+            return ERANGE;
+        }
+        conf.kadmind_port = htons(n);
+        conf.mask |= KADM5_CONFIG_KADMIND_PORT;
+        server = s;
+    }
+    if (server) {
+        conf.admin_server = s ? strdup(server) : strdup(server);
+        conf.mask |= KADM5_CONFIG_ADMIN_SERVER;
+        free(s);
+        s = NULL;
+
+        if (conf.admin_server == NULL)
+            err(1, "Out of memory");
+    } else if (hdb) {
+        conf.dbname = strdup(hdb);
+        conf.mask |= KADM5_CONFIG_DBNAME;
+        if (conf.dbname == NULL)
+            err(1, "Out of memory");
+    }
+    if (realm) {
+        krb5_set_default_realm(context, realm); /* XXX ??? */
```

Comment:
Let's not use this.  Instead we should parse princ names from HTTP requests using `krb5_parse_name_flags()` and then set the realm if it wasn't.